draft-ietf-acme-acme.md

```diff
@@ -522,7 +522,9 @@ way as a value it had never issued.
 When a server rejects a request because its nonce value was unacceptable (or not
 present), it MUST provide HTTP status code 400 (Bad Request), and indicate the
 ACME error type "urn:ietf:params:acme:error:badNonce".  An error response with
-the "badNonce" error type MUST include a Replay-Nonce header with a fresh nonce.
+the "badNonce" error type MUST include a Replay-Nonce header with a
+fresh nonce that the server will accept in a retry of the original
+query (and possibly in other requests).
```

Comment:
I hope I'm not too late for discussion... But what is "and possibly in other requests" supposed to mean? Does it mean that a client MUST NOT or SHOULD NOT use the Replay-Nonce from the badNonce error for other requests?

I'm not happy with this idea. Until now, nonces were not bound to a certain type of request. A client could just remember the nonce it received from the last request, and use it for the next request. If I take this change literally, a client would now have to store multiple nonces, and find the one matching for a certain request. I see no advantage or increase of security in that.

Or did I just misunderstand it?

Reply:
You've misunderstood a little, but in a totally understandable way :)  Right now, ACME doesn't say anything about how nonces are scoped -- in principle, the server could scope them per-client, per-request-type, etc.  There are valid reasons for servers to do some of these things, mainly related to managing server state.

My intent with this text was to specify the minimum necessary scope constraint for a client and server to interoperate.  Without this requirement, the worst case is that the server has some crazy nonce management strategy and client can never send a valid request.  The requirement for the nonce to be valid on retry assures that the worst case is that the client has to do every request twice.  Obviously, I hope that most servers will use more sensible strategies that don't cause this condition!

I agree that the current text is awkward, though.  Would it be helpful to replace that parenthetical with something like the following?

> Other than this constraint, ACME does not constrain how servers scope nonces.  Clients are MAY assume that nonces have broad scope, e.g., by having a single pool of nonces used for all requests.  However, when retrying in response to a "badNonce" error, the client MUST use the nonce provided in the error response.   Obviously, servers should scope nonces broadly enough that retries are not needed very often.

Reply:
I think Richard's proposed text here is good and addresses the review comment. @richsalz, are you good with this? I think it's the last thing.